src/vm/dllimport.cpp

```diff
@@ -6342,11 +6343,81 @@ NATIVE_LIBRARY_HANDLE NDirect::LoadLibraryModuleViaHost(NDirectMethodDesc * pMD,
     args[ARGNUM_1]  = PTR_TO_ARGHOLDER(ptrManagedAssemblyLoadContext);
 
     // Make the call
-    CALL_MANAGED_METHOD(hmod,LPVOID,args);
+    CALL_MANAGED_METHOD(hmod, NATIVE_LIBRARY_HANDLE, args);
 
     GCPROTECT_END();
 
-    return (NATIVE_LIBRARY_HANDLE)hmod;
+    return hmod;
+}
+
+// Return the AssemblyLoadContext for an assembly
+INT_PTR GetManagedAssemblyLoadContext(Assembly* pAssembly)
+{
+    STANDARD_VM_CONTRACT;
+
+    PTR_ICLRPrivBinder pBindingContext = pAssembly->GetManifestFile()->GetBindingContext();
+    if (pBindingContext == NULL)
+    {
+        // GetBindingContext() returns NULL for System.Private.CoreLib
+        return NULL;
+    }
+
+    UINT_PTR assemblyBinderID = 0;
+    IfFailThrow(pBindingContext->GetBinderID(&assemblyBinderID));
+
+    ICLRPrivBinder *pCurrentBinder = reinterpret_cast<ICLRPrivBinder *>(assemblyBinderID);
+    ICLRPrivBinder *pTPABinder = GetAppDomain()->GetTPABinderContext();
+    BOOL isDefaultContext = AreSameBinderInstance(pCurrentBinder, pTPABinder);
+
+    INT_PTR ptrManagedAssemblyLoadContext = (isDefaultContext) ?
+        ((CLRPrivBinderCoreCLR *)pCurrentBinder)->GetManagedAssemblyLoadContext() :
+        ((CLRPrivBinderAssemblyLoadContext *)pCurrentBinder)->GetManagedAssemblyLoadContext();
+
+    return ptrManagedAssemblyLoadContext;
+}
+
+// static
+NATIVE_LIBRARY_HANDLE NDirect::LoadLibraryModuleViaEvent(NDirectMethodDesc * pMD, PCWSTR wszLibName)
+{
+    STANDARD_VM_CONTRACT;
```

Comment:
Shouldn't this contract indicate the GC mode as well as the fact that it can trigger a GC? I don't know what the "standard" contract is, sorry.

Reply:
`STANDARD_VM_CONTRACT` is a short-cut for `THROWS; GC_TRIGGERS; MODE_PREEMPTIVE;`. It means ordinary code without special constrains (ie same constrains as arbitrary 3rd party code). `STANDARD_VM_CONTRACT` and `LIMITED_METHOD_CONTRACT` are the preferred contracts to use since they do not come with any magic.